---
output: html_document
---


ENCODEdb: A compilation of metadata from ENCODE
====================================================================

Audrey Lema√ßon, Charles Joly Beauparlant and Arnaud Droit.

This package and the underlying ENCODEdb code are distributed under 
the Artistic license 2.0. You are free to use and redistribute this software. 

"The ENCODE (Encyclopedia of DNA Elements) Consortium is an international 
collaboration of research groups funded by the National Human Genome Research 
Institute (NHGRI). The goal of ENCODE is to build a comprehensive parts list of 
functional elements in the human genome, including elements that act at the 
protein and RNA levels, and regulatory elements that control cells and 
circumstances in which a gene is active"[^1].

However, data retrieval and downloading can be really time consuming using 
current web portal. 

This package has been designed to facilitate the data access by compiling the 
metadata associated with file, experiment, dataset, biosample, and treatment. 

We first extract ENCODE schema from its public github repository to rebuild 
the ENCODE database into an SQLite database. Thanks to this package, the user 
will be enable to generate, store and query ENCODE database locally. We also 
developped a function which can extract the essential metadata in a R object to 
aid data exploration.
We implemented time-saving features to select ENCODE files by querying their 
metadata and download them.

The SQLite database can be regenerated at will to keep it up-to-date.

This vignette will introduce all the main features of the ENCODEdb package.

### Loading ENCODEdb package

```{r libraryLoad}
suppressMessages(library(ENCODEdb))
```

### Introduction

This package comes with a up-to-date list of dataframe containing the essential 
of ENCODE files metadata: `encode_df`. This list contains two elements.
The first one `encode_df$experiment` is a data.frame containing essential 
information for each file part of an experiment ; the second one `encode_df$dataset`
is a data.frame containing essential information for each file part of a dataset.

This `encode_df` is **mandatory** to use the functions provided in this package.

### Main functions

#### Query

The `query` function allow the user to find the subset of files corresponding to
a precise query define following 10 criteria :

|parameter| description|available for|
|---------|------------|-------------|
|accession|*the experiment or dataset accession*|experiment / dataset|
|assay|*the assay type*|experiment|
|biosample|*the biosample name*|experiment|
|dataset_access[^2]|*the dataset accession*|experiment / dataset|
|file_accession|*the file accesion*|experiment / dataset|
|file_format|*the file format*|experiment / dataset|
|lab|*the laboratory*|experiment / dataset|
|organism|*the donor organism*|experiment|
|target|*the experimental target*|experiment|
|treatment|*the treatment*|experiment|


By default, the query use the exact string matching to perform the selection of 
the relevant entries. This behavior can be changed by setting the `fixed` option 
to **FALSE**.

The structure of the result set is similar to the `encode_df` structure : a list 
of two elements *experiment* and *dataset*.

For example, to select fastq files produced by RNA-seq assay on human cell MCF-7 :
```{r query_results, collapse=TRUE}
query_results <- query(assay = "RNA-seq", organism = "Homo sapiens", biosample = "MCF-7", file_format = "fastq", fixed = T)
```

The same request with approximate spelling of the assay type and `fixed` option to **TRUE**, will give no results :
```{r query_results_2, collapse=TRUE}
query_results <- query(assay = "rnaseq", organism = "Homo sapiens", biosample = "MCF-7", file_format = "fastq", fixed = T)
```

If you follow the warning guidance and set the `fixed` option to **FALSE** : 
```{r query_results_3, collapse=TRUE}
query_results <- query(assay = "rnaseq", organism = "Homo sapiens", biosample = "MCF-7", file_format = "fastq", fixed = F)
```


These criteria correspond to the filters that you can find on ENCODE portal : 

![results of a filtered search on ENCODE portal](img/query_mcf7.png)

*Note : the usage of some criteria, like organism or target, will automatically 
dismissed the dataset results because this information isn't available for that type of data.*


#### Search
This function simulates a key word search that you could perform through the 
ENCODE web portal.

The `search` function return a data frame which corresponds to the result page 
provided by ENCODE portal.

Here is the example of the following search : *"a549 chip-seq homo sapiens"*.


On ENCODE portal :

![results of a key word search on ENCODE portal](img/search_a549.png)

With our function :
```{r search_results, collapse=TRUE}
search_results <- search(searchTerm = "a549 chip-seq homo sapiens", limit = "all")
```


#### Download

Following a search or a query, you may want to download the corresponding files.
When using, the ENCODE portal you have to select the experiments one after an 
other and for each experiment, select manually the files you're interested in.

Our `download` function is a real time saving feature. To use it, you have to 
provide the results set that you've just get from the `search` or `query` 
function, indicate the origin of the dataset ("search" or "query") and then the 
path to the directory where you want to copy the downloaded files (default "/tmp").

To ensure that the downloading have succeeded, we conduct a check md5 sum 
comparison for each file.

Moreover, if your results set stem from the `search` function, you may want to 
restrict the downloading to a certain type of file. To do so, you can set the 
`format` option ( by default set to "all")

Here is a small query :

```{r query_results_4, collapse=TRUE}
query_results <- query(assay = "switchgear", target ="elavl1", fixed = F)
```

And its equivalent search
```{r search_results_2, collapse=TRUE}
search_results <- search(searchTerm = "switchgear elavl1", limit = "all")
```

To select a particular file format you can :

1/ add this filter to your query and then run the `download` function
```{r query_results_5, collapse=TRUE}
query_results <- query(assay = "switchgear", target ="elavl1", file_format = "bed_broadPeak" , fixed = F)
download(resultSet = query_results, resultOrigin = "query")
```

2/ specify the format to the `download` function
```{r collapse=TRUE}
download(resultSet = search_results, resultOrigin = "search", format = "bed_broadPeak")
```

### Data update

If you want regenerate this list, you have to process the following steps :

* generate the SQL database from ENCODE 

```{r tables, eval=FALSE}
# the path (relative or absolute) to the future database
database_filename = "encode.sqlite"
tables = prepare_ENCODEdb(database_filename)
```

* generate the metadata encode_df from the SQL database
```{r encode_df, eval = FALSE}
encode_df <- export_ENCODEdb_matrix(database_filename)
```


The whole process will take several minutes (5 to 15 minutes depending on your
work environment)

You can also use the SQL database for own purpose. The imputed database 
model is available in XXXXXXXXX

[^1]: source : [ENCODE Projet Portal ](https://www.encodeproject.org/)

[^2]:There is a subtle difference between the parameters **accession** and 
**dataset_accession**. In fact, some files can be part of experiment, dataset or 
both.
When using **accession**, you will get all the files directly link to this accession 
(experiment or dataset). While the usage of **dataset_accesstion** will get the 
files directly link to the requested dataset **AND** those which are part of an 
experiment and indirectly link to a dataset (reported as related_files in the 
dataset and related_dataset in experiment)
